Gradle 7.6.3 fails with plugins that use multi-release JAR files

### Current Behavior

If I apply a plugin that uses H2 ('com.h2database:h2:2.2.224'), a multi-release JAR, the execution of the plugin fails.

```groovy
plugins {
  id("org.owasp.dependencycheck") version "9.0.0"
}
```

I then run `./gradlew dependencyCheckAnalyze` I get the following error/stacktrace:

```bash
FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring project ':lib'.
> java.util.concurrent.ExecutionException: org.gradle.api.GradleException: Failed to create Jar file /Users/jeremy/.gradle/caches/jars-9/29a26170a4fefd5316fa39bfac1c4ff4/h2-2.2.224.jar.

* Try:
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Exception is:
org.gradle.api.ProjectConfigurationException: A problem occurred configuring project ':lib'.
        at org.gradle.configuration.project.LifecycleProjectEvaluator.wrapException(LifecycleProjectEvaluator.java:84)
        at org.gradle.configuration.project.LifecycleProjectEvaluator.addConfigurationFailure(LifecycleProjectEvaluator.java:77)
        at org.gradle.configuration.project.LifecycleProjectEvaluator.access$400(LifecycleProjectEvaluator.java:55)
        at org.gradle.configuration.project.LifecycleProjectEvaluator$EvaluateProject.lambda$run$0(LifecycleProjectEvaluator.java:111)
        at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.lambda$applyToMutableState$0(DefaultProjectStateRegistry.java:360)
        at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.lambda$fromMutableState$1(DefaultProjectStateRegistry.java:383)
        at org.gradle.internal.work.DefaultWorkerLeaseService.withReplacedLocks(DefaultWorkerLeaseService.java:345)
        at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.fromMutableState(DefaultProjectStateRegistry.java:383)
        at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.applyToMutableState(DefaultProjectStateRegistry.java:359)
        at org.gradle.configuration.project.LifecycleProjectEvaluator$EvaluateProject.run(LifecycleProjectEvaluator.java:100)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:47)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:68)
        at org.gradle.configuration.project.LifecycleProjectEvaluator.evaluate(LifecycleProjectEvaluator.java:72)
        at org.gradle.api.internal.project.DefaultProject.evaluate(DefaultProject.java:762)
        at org.gradle.api.internal.project.DefaultProject.evaluate(DefaultProject.java:153)
        at org.gradle.api.internal.project.ProjectLifecycleController.lambda$ensureSelfConfigured$1(ProjectLifecycleController.java:63)
        at org.gradle.internal.model.StateTransitionController.lambda$doTransition$12(StateTransitionController.java:236)
        at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:247)
        at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:235)
        at org.gradle.internal.model.StateTransitionController.lambda$maybeTransitionIfNotCurrentlyTransitioning$9(StateTransitionController.java:196)
        at org.gradle.internal.work.DefaultSynchronizer.withLock(DefaultSynchronizer.java:34)
        at org.gradle.internal.model.StateTransitionController.maybeTransitionIfNotCurrentlyTransitioning(StateTransitionController.java:192)
        at org.gradle.api.internal.project.ProjectLifecycleController.ensureSelfConfigured(ProjectLifecycleController.java:63)
        at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.ensureConfigured(DefaultProjectStateRegistry.java:334)
        at org.gradle.execution.TaskPathProjectEvaluator.configure(TaskPathProjectEvaluator.java:33)
        at org.gradle.execution.TaskPathProjectEvaluator.configureHierarchy(TaskPathProjectEvaluator.java:49)
        at org.gradle.configuration.DefaultProjectsPreparer.prepareProjects(DefaultProjectsPreparer.java:50)
        at org.gradle.configuration.BuildTreePreparingProjectsPreparer.prepareProjects(BuildTreePreparingProjectsPreparer.java:64)
        at org.gradle.configuration.BuildOperationFiringProjectsPreparer$ConfigureBuild.run(BuildOperationFiringProjectsPreparer.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:47)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:68)
        at org.gradle.configuration.BuildOperationFiringProjectsPreparer.prepareProjects(BuildOperationFiringProjectsPreparer.java:40)
        at org.gradle.initialization.VintageBuildModelController.lambda$prepareProjects$2(VintageBuildModelController.java:84)
        at org.gradle.internal.model.StateTransitionController.lambda$doTransition$12(StateTransitionController.java:236)
        at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:247)
        at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:235)
        at org.gradle.internal.model.StateTransitionController.lambda$transitionIfNotPreviously$10(StateTransitionController.java:210)
        at org.gradle.internal.work.DefaultSynchronizer.withLock(DefaultSynchronizer.java:34)
        at org.gradle.internal.model.StateTransitionController.transitionIfNotPreviously(StateTransitionController.java:206)
        at org.gradle.initialization.VintageBuildModelController.prepareProjects(VintageBuildModelController.java:84)
        at org.gradle.initialization.VintageBuildModelController.prepareToScheduleTasks(VintageBuildModelController.java:71)
        at org.gradle.internal.build.DefaultBuildLifecycleController.lambda$prepareToScheduleTasks$2(DefaultBuildLifecycleController.java:129)
        at org.gradle.internal.model.StateTransitionController.lambda$doTransition$12(StateTransitionController.java:236)
        at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:247)
        at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:235)
        at org.gradle.internal.model.StateTransitionController.lambda$maybeTransition$8(StateTransitionController.java:187)
        at org.gradle.internal.work.DefaultSynchronizer.withLock(DefaultSynchronizer.java:34)
        at org.gradle.internal.model.StateTransitionController.maybeTransition(StateTransitionController.java:183)
        at org.gradle.internal.build.DefaultBuildLifecycleController.prepareToScheduleTasks(DefaultBuildLifecycleController.java:127)
        at org.gradle.internal.buildtree.DefaultBuildTreeWorkPreparer.scheduleRequestedTasks(DefaultBuildTreeWorkPreparer.java:36)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.lambda$doScheduleAndRunTasks$3(DefaultBuildTreeLifecycleController.java:96)
        at org.gradle.composite.internal.DefaultIncludedBuildTaskGraph.withNewWorkGraph(DefaultIncludedBuildTaskGraph.java:109)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.doScheduleAndRunTasks(DefaultBuildTreeLifecycleController.java:95)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.lambda$scheduleAndRunTasks$1(DefaultBuildTreeLifecycleController.java:76)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.lambda$runBuild$5(DefaultBuildTreeLifecycleController.java:113)
        at org.gradle.internal.model.StateTransitionController.lambda$transition$5(StateTransitionController.java:166)
        at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:247)
        at org.gradle.internal.model.StateTransitionController.lambda$transition$6(StateTransitionController.java:166)
        at org.gradle.internal.work.DefaultSynchronizer.withLock(DefaultSynchronizer.java:44)
        at org.gradle.internal.model.StateTransitionController.transition(StateTransitionController.java:166)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.runBuild(DefaultBuildTreeLifecycleController.java:110)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.scheduleAndRunTasks(DefaultBuildTreeLifecycleController.java:76)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.scheduleAndRunTasks(DefaultBuildTreeLifecycleController.java:71)
        at org.gradle.tooling.internal.provider.ExecuteBuildActionRunner.run(ExecuteBuildActionRunner.java:31)
        at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
        at org.gradle.internal.buildtree.ProblemReportingBuildActionRunner.run(ProblemReportingBuildActionRunner.java:49)
        at org.gradle.launcher.exec.BuildOutcomeReportingBuildActionRunner.run(BuildOutcomeReportingBuildActionRunner.java:65)
        at org.gradle.tooling.internal.provider.FileSystemWatchingBuildActionRunner.run(FileSystemWatchingBuildActionRunner.java:136)
        at org.gradle.launcher.exec.BuildCompletionNotifyingBuildActionRunner.run(BuildCompletionNotifyingBuildActionRunner.java:41)
        at org.gradle.launcher.exec.RootBuildLifecycleBuildActionExecutor.lambda$execute$0(RootBuildLifecycleBuildActionExecutor.java:40)
        at org.gradle.composite.internal.DefaultRootBuildState.run(DefaultRootBuildState.java:122)
        at org.gradle.launcher.exec.RootBuildLifecycleBuildActionExecutor.execute(RootBuildLifecycleBuildActionExecutor.java:40)
        at org.gradle.internal.buildtree.DefaultBuildTreeContext.execute(DefaultBuildTreeContext.java:40)
        at org.gradle.launcher.exec.BuildTreeLifecycleBuildActionExecutor.lambda$execute$0(BuildTreeLifecycleBuildActionExecutor.java:65)
        at org.gradle.internal.buildtree.BuildTreeState.run(BuildTreeState.java:53)
        at org.gradle.launcher.exec.BuildTreeLifecycleBuildActionExecutor.execute(BuildTreeLifecycleBuildActionExecutor.java:65)
        at org.gradle.launcher.exec.RunAsBuildOperationBuildActionExecutor$3.call(RunAsBuildOperationBuildActionExecutor.java:61)
        at org.gradle.launcher.exec.RunAsBuildOperationBuildActionExecutor$3.call(RunAsBuildOperationBuildActionExecutor.java:57)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:204)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:199)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:73)
        at org.gradle.launcher.exec.RunAsBuildOperationBuildActionExecutor.execute(RunAsBuildOperationBuildActionExecutor.java:57)
        at org.gradle.launcher.exec.RunAsWorkerThreadBuildActionExecutor.lambda$execute$0(RunAsWorkerThreadBuildActionExecutor.java:36)
        at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:249)
        at org.gradle.internal.work.DefaultWorkerLeaseService.runAsWorkerThread(DefaultWorkerLeaseService.java:109)
        at org.gradle.launcher.exec.RunAsWorkerThreadBuildActionExecutor.execute(RunAsWorkerThreadBuildActionExecutor.java:36)
        at org.gradle.tooling.internal.provider.continuous.ContinuousBuildActionExecutor.execute(ContinuousBuildActionExecutor.java:110)
        at org.gradle.tooling.internal.provider.SubscribableBuildActionExecutor.execute(SubscribableBuildActionExecutor.java:64)
        at org.gradle.internal.session.DefaultBuildSessionContext.execute(DefaultBuildSessionContext.java:46)
        at org.gradle.tooling.internal.provider.BuildSessionLifecycleBuildActionExecuter$ActionImpl.apply(BuildSessionLifecycleBuildActionExecuter.java:100)
        at org.gradle.tooling.internal.provider.BuildSessionLifecycleBuildActionExecuter$ActionImpl.apply(BuildSessionLifecycleBuildActionExecuter.java:88)
        at org.gradle.internal.session.BuildSessionState.run(BuildSessionState.java:69)
        at org.gradle.tooling.internal.provider.BuildSessionLifecycleBuildActionExecuter.execute(BuildSessionLifecycleBuildActionExecuter.java:62)
        at org.gradle.tooling.internal.provider.BuildSessionLifecycleBuildActionExecuter.execute(BuildSessionLifecycleBuildActionExecuter.java:41)
        at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:63)
        at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:31)
        at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:52)
        at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:40)
        at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:47)
        at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:31)
        at org.gradle.launcher.daemon.server.exec.ExecuteBuild.doBuild(ExecuteBuild.java:65)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.WatchForDisconnection.execute(WatchForDisconnection.java:39)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.ResetDeprecationLogger.execute(ResetDeprecationLogger.java:29)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.RequestStopIfSingleUsedDaemon.execute(RequestStopIfSingleUsedDaemon.java:35)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.create(ForwardClientInput.java:78)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.create(ForwardClientInput.java:75)
        at org.gradle.util.internal.Swapper.swap(Swapper.java:38)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput.execute(ForwardClientInput.java:75)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.LogAndCheckHealth.execute(LogAndCheckHealth.java:55)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.LogToClient.doBuild(LogToClient.java:63)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.EstablishBuildEnvironment.doBuild(EstablishBuildEnvironment.java:84)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.StartBuildOrRespondWithBusy$1.run(StartBuildOrRespondWithBusy.java:52)
        at org.gradle.launcher.daemon.server.DaemonStateCoordinator$1.run(DaemonStateCoordinator.java:297)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:49)
Caused by: org.gradle.internal.UncheckedException: java.util.concurrent.ExecutionException: org.gradle.api.GradleException: Failed to create Jar file /Users/jeremy/.gradle/caches/jars-9/29a26170a4fefd5316fa39bfac1c4ff4/h2-2.2.224.jar.
        at org.gradle.internal.UncheckedException.throwAsUncheckedException(UncheckedException.java:68)
        at org.gradle.internal.UncheckedException.throwAsUncheckedException(UncheckedException.java:41)
        at org.gradle.internal.UncheckedException.unchecked(UncheckedException.java:76)
        at org.gradle.internal.classpath.DefaultCachedClasspathTransformer.lambda$transformAll$13(DefaultCachedClasspathTransformer.java:244)
        at org.gradle.cache.internal.DefaultCacheAccess.useCache(DefaultCacheAccess.java:220)
        at org.gradle.cache.internal.DefaultPersistentDirectoryStore.useCache(DefaultPersistentDirectoryStore.java:198)
        at org.gradle.cache.internal.DefaultCacheFactory$ReferenceTrackingCache.useCache(DefaultCacheFactory.java:219)
        at org.gradle.internal.classpath.DefaultCachedClasspathTransformer.transformAll(DefaultCachedClasspathTransformer.java:220)
        at org.gradle.internal.classpath.DefaultCachedClasspathTransformer.transformFiles(DefaultCachedClasspathTransformer.java:127)
        at org.gradle.internal.classpath.DefaultCachedClasspathTransformer.transform(DefaultCachedClasspathTransformer.java:94)
        at org.gradle.plugin.use.internal.DefaultPluginRequestApplicator.exportBuildLogicClassPathTo(DefaultPluginRequestApplicator.java:187)
        at org.gradle.plugin.use.internal.DefaultPluginRequestApplicator.defineScriptHandlerClassScope(DefaultPluginRequestApplicator.java:177)
        at org.gradle.plugin.use.internal.DefaultPluginRequestApplicator.applyPlugins(DefaultPluginRequestApplicator.java:142)
        at org.gradle.configuration.DefaultScriptPluginFactory$ScriptPluginImpl.apply(DefaultScriptPluginFactory.java:117)
        at org.gradle.configuration.BuildOperationScriptPlugin$1.run(BuildOperationScriptPlugin.java:65)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:47)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:68)
        at org.gradle.configuration.BuildOperationScriptPlugin.lambda$apply$0(BuildOperationScriptPlugin.java:62)
        at org.gradle.configuration.internal.DefaultUserCodeApplicationContext.apply(DefaultUserCodeApplicationContext.java:44)
        at org.gradle.configuration.BuildOperationScriptPlugin.apply(BuildOperationScriptPlugin.java:62)
        at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.lambda$applyToMutableState$0(DefaultProjectStateRegistry.java:360)
        at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.fromMutableState(DefaultProjectStateRegistry.java:378)
        at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.applyToMutableState(DefaultProjectStateRegistry.java:359)
        at org.gradle.configuration.project.BuildScriptProcessor.execute(BuildScriptProcessor.java:42)
        at org.gradle.configuration.project.BuildScriptProcessor.execute(BuildScriptProcessor.java:26)
        at org.gradle.configuration.project.ConfigureActionsProjectEvaluator.evaluate(ConfigureActionsProjectEvaluator.java:35)
        at org.gradle.configuration.project.LifecycleProjectEvaluator$EvaluateProject.lambda$run$0(LifecycleProjectEvaluator.java:109)
        at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.lambda$applyToMutableState$0(DefaultProjectStateRegistry.java:360)
        at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.lambda$fromMutableState$1(DefaultProjectStateRegistry.java:383)
        at org.gradle.internal.work.DefaultWorkerLeaseService.withReplacedLocks(DefaultWorkerLeaseService.java:345)
        at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.fromMutableState(DefaultProjectStateRegistry.java:383)
        at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.applyToMutableState(DefaultProjectStateRegistry.java:359)
        at org.gradle.configuration.project.LifecycleProjectEvaluator$EvaluateProject.run(LifecycleProjectEvaluator.java:100)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:47)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:68)
        at org.gradle.configuration.project.LifecycleProjectEvaluator.evaluate(LifecycleProjectEvaluator.java:72)
        at org.gradle.api.internal.project.DefaultProject.evaluate(DefaultProject.java:762)
        at org.gradle.api.internal.project.DefaultProject.evaluate(DefaultProject.java:153)
        at org.gradle.api.internal.project.ProjectLifecycleController.lambda$ensureSelfConfigured$1(ProjectLifecycleController.java:63)
        at org.gradle.internal.model.StateTransitionController.lambda$doTransition$12(StateTransitionController.java:236)
        at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:247)
        at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:235)
        at org.gradle.internal.model.StateTransitionController.lambda$maybeTransitionIfNotCurrentlyTransitioning$9(StateTransitionController.java:196)
        at org.gradle.internal.work.DefaultSynchronizer.withLock(DefaultSynchronizer.java:34)
        at org.gradle.internal.model.StateTransitionController.maybeTransitionIfNotCurrentlyTransitioning(StateTransitionController.java:192)
        at org.gradle.api.internal.project.ProjectLifecycleController.ensureSelfConfigured(ProjectLifecycleController.java:63)
        at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.ensureConfigured(DefaultProjectStateRegistry.java:334)
        at org.gradle.execution.TaskPathProjectEvaluator.configure(TaskPathProjectEvaluator.java:33)
        at org.gradle.execution.TaskPathProjectEvaluator.configureHierarchy(TaskPathProjectEvaluator.java:49)
        at org.gradle.configuration.DefaultProjectsPreparer.prepareProjects(DefaultProjectsPreparer.java:50)
        at org.gradle.configuration.BuildTreePreparingProjectsPreparer.prepareProjects(BuildTreePreparingProjectsPreparer.java:64)
        at org.gradle.configuration.BuildOperationFiringProjectsPreparer$ConfigureBuild.run(BuildOperationFiringProjectsPreparer.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:47)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:68)
        at org.gradle.configuration.BuildOperationFiringProjectsPreparer.prepareProjects(BuildOperationFiringProjectsPreparer.java:40)
        at org.gradle.initialization.VintageBuildModelController.lambda$prepareProjects$2(VintageBuildModelController.java:84)
        at org.gradle.internal.model.StateTransitionController.lambda$doTransition$12(StateTransitionController.java:236)
        at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:247)
        at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:235)
        at org.gradle.internal.model.StateTransitionController.lambda$transitionIfNotPreviously$10(StateTransitionController.java:210)
        at org.gradle.internal.work.DefaultSynchronizer.withLock(DefaultSynchronizer.java:34)
        at org.gradle.internal.model.StateTransitionController.transitionIfNotPreviously(StateTransitionController.java:206)
        at org.gradle.initialization.VintageBuildModelController.prepareProjects(VintageBuildModelController.java:84)
        at org.gradle.initialization.VintageBuildModelController.prepareToScheduleTasks(VintageBuildModelController.java:71)
        at org.gradle.internal.build.DefaultBuildLifecycleController.lambda$prepareToScheduleTasks$2(DefaultBuildLifecycleController.java:129)
        at org.gradle.internal.model.StateTransitionController.lambda$doTransition$12(StateTransitionController.java:236)
        at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:247)
        at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:235)
        at org.gradle.internal.model.StateTransitionController.lambda$maybeTransition$8(StateTransitionController.java:187)
        at org.gradle.internal.work.DefaultSynchronizer.withLock(DefaultSynchronizer.java:34)
        at org.gradle.internal.model.StateTransitionController.maybeTransition(StateTransitionController.java:183)
        at org.gradle.internal.build.DefaultBuildLifecycleController.prepareToScheduleTasks(DefaultBuildLifecycleController.java:127)
        at org.gradle.internal.buildtree.DefaultBuildTreeWorkPreparer.scheduleRequestedTasks(DefaultBuildTreeWorkPreparer.java:36)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.lambda$doScheduleAndRunTasks$3(DefaultBuildTreeLifecycleController.java:96)
        at org.gradle.composite.internal.DefaultIncludedBuildTaskGraph.withNewWorkGraph(DefaultIncludedBuildTaskGraph.java:109)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.doScheduleAndRunTasks(DefaultBuildTreeLifecycleController.java:95)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.lambda$scheduleAndRunTasks$1(DefaultBuildTreeLifecycleController.java:76)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.lambda$runBuild$5(DefaultBuildTreeLifecycleController.java:113)
        at org.gradle.internal.model.StateTransitionController.lambda$transition$5(StateTransitionController.java:166)
        at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:247)
        at org.gradle.internal.model.StateTransitionController.lambda$transition$6(StateTransitionController.java:166)
        at org.gradle.internal.work.DefaultSynchronizer.withLock(DefaultSynchronizer.java:44)
        at org.gradle.internal.model.StateTransitionController.transition(StateTransitionController.java:166)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.runBuild(DefaultBuildTreeLifecycleController.java:110)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.scheduleAndRunTasks(DefaultBuildTreeLifecycleController.java:76)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.scheduleAndRunTasks(DefaultBuildTreeLifecycleController.java:71)
        at org.gradle.tooling.internal.provider.ExecuteBuildActionRunner.run(ExecuteBuildActionRunner.java:31)
        at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
        at org.gradle.internal.buildtree.ProblemReportingBuildActionRunner.run(ProblemReportingBuildActionRunner.java:49)
        at org.gradle.launcher.exec.BuildOutcomeReportingBuildActionRunner.run(BuildOutcomeReportingBuildActionRunner.java:65)
        at org.gradle.tooling.internal.provider.FileSystemWatchingBuildActionRunner.run(FileSystemWatchingBuildActionRunner.java:136)
        at org.gradle.launcher.exec.BuildCompletionNotifyingBuildActionRunner.run(BuildCompletionNotifyingBuildActionRunner.java:41)
        at org.gradle.launcher.exec.RootBuildLifecycleBuildActionExecutor.lambda$execute$0(RootBuildLifecycleBuildActionExecutor.java:40)
        at org.gradle.composite.internal.DefaultRootBuildState.run(DefaultRootBuildState.java:122)
        at org.gradle.launcher.exec.RootBuildLifecycleBuildActionExecutor.execute(RootBuildLifecycleBuildActionExecutor.java:40)
        at org.gradle.internal.buildtree.DefaultBuildTreeContext.execute(DefaultBuildTreeContext.java:40)
        at org.gradle.launcher.exec.BuildTreeLifecycleBuildActionExecutor.lambda$execute$0(BuildTreeLifecycleBuildActionExecutor.java:65)
        at org.gradle.internal.buildtree.BuildTreeState.run(BuildTreeState.java:53)
        at org.gradle.launcher.exec.BuildTreeLifecycleBuildActionExecutor.execute(BuildTreeLifecycleBuildActionExecutor.java:65)
        at org.gradle.launcher.exec.RunAsBuildOperationBuildActionExecutor$3.call(RunAsBuildOperationBuildActionExecutor.java:61)
        at org.gradle.launcher.exec.RunAsBuildOperationBuildActionExecutor$3.call(RunAsBuildOperationBuildActionExecutor.java:57)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:204)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:199)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:73)
        at org.gradle.launcher.exec.RunAsBuildOperationBuildActionExecutor.execute(RunAsBuildOperationBuildActionExecutor.java:57)
        at org.gradle.launcher.exec.RunAsWorkerThreadBuildActionExecutor.lambda$execute$0(RunAsWorkerThreadBuildActionExecutor.java:36)
        at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:249)
        at org.gradle.internal.work.DefaultWorkerLeaseService.runAsWorkerThread(DefaultWorkerLeaseService.java:109)
        at org.gradle.launcher.exec.RunAsWorkerThreadBuildActionExecutor.execute(RunAsWorkerThreadBuildActionExecutor.java:36)
        at org.gradle.tooling.internal.provider.continuous.ContinuousBuildActionExecutor.execute(ContinuousBuildActionExecutor.java:110)
        at org.gradle.tooling.internal.provider.SubscribableBuildActionExecutor.execute(SubscribableBuildActionExecutor.java:64)
        at org.gradle.internal.session.DefaultBuildSessionContext.execute(DefaultBuildSessionContext.java:46)
        at org.gradle.tooling.internal.provider.BuildSessionLifecycleBuildActionExecuter$ActionImpl.apply(BuildSessionLifecycleBuildActionExecuter.java:100)
        at org.gradle.tooling.internal.provider.BuildSessionLifecycleBuildActionExecuter$ActionImpl.apply(BuildSessionLifecycleBuildActionExecuter.java:88)
        at org.gradle.internal.session.BuildSessionState.run(BuildSessionState.java:69)
        at org.gradle.tooling.internal.provider.BuildSessionLifecycleBuildActionExecuter.execute(BuildSessionLifecycleBuildActionExecuter.java:62)
        at org.gradle.tooling.internal.provider.BuildSessionLifecycleBuildActionExecuter.execute(BuildSessionLifecycleBuildActionExecuter.java:41)
        at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:63)
        at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:31)
        at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:52)
        at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:40)
        at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:47)
        at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:31)
        at org.gradle.launcher.daemon.server.exec.ExecuteBuild.doBuild(ExecuteBuild.java:65)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.WatchForDisconnection.execute(WatchForDisconnection.java:39)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.ResetDeprecationLogger.execute(ResetDeprecationLogger.java:29)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.RequestStopIfSingleUsedDaemon.execute(RequestStopIfSingleUsedDaemon.java:35)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.create(ForwardClientInput.java:78)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.create(ForwardClientInput.java:75)
        at org.gradle.util.internal.Swapper.swap(Swapper.java:38)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput.execute(ForwardClientInput.java:75)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.LogAndCheckHealth.execute(LogAndCheckHealth.java:55)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.LogToClient.doBuild(LogToClient.java:63)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.EstablishBuildEnvironment.doBuild(EstablishBuildEnvironment.java:84)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.StartBuildOrRespondWithBusy$1.run(StartBuildOrRespondWithBusy.java:52)
        at org.gradle.launcher.daemon.server.DaemonStateCoordinator$1.run(DaemonStateCoordinator.java:297)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:49)
Caused by: java.util.concurrent.ExecutionException: org.gradle.api.GradleException: Failed to create Jar file /Users/jeremy/.gradle/caches/jars-9/29a26170a4fefd5316fa39bfac1c4ff4/h2-2.2.224.jar.
        at org.gradle.internal.UncheckedException.unchecked(UncheckedException.java:74)
        ... 169 more
Caused by: org.gradle.api.GradleException: Failed to create Jar file /Users/jeremy/.gradle/caches/jars-9/29a26170a4fefd5316fa39bfac1c4ff4/h2-2.2.224.jar.
        at org.gradle.internal.classpath.ClasspathBuilder.jar(ClasspathBuilder.java:57)
        at org.gradle.internal.classpath.InstrumentingClasspathFileTransformer.instrument(InstrumentingClasspathFileTransformer.java:137)
        at org.gradle.internal.classpath.InstrumentingClasspathFileTransformer.transform(InstrumentingClasspathFileTransformer.java:132)
        at org.gradle.internal.classpath.InstrumentingClasspathFileTransformer.transform(InstrumentingClasspathFileTransformer.java:96)
        at org.gradle.internal.classpath.DefaultCachedClasspathTransformer.transformFile(DefaultCachedClasspathTransformer.java:193)
        at org.gradle.internal.classpath.DefaultCachedClasspathTransformer.lambda$cachedFile$6(DefaultCachedClasspathTransformer.java:186)
        at org.gradle.internal.UncheckedException.unchecked(UncheckedException.java:74)
        at org.gradle.internal.classpath.DefaultCachedClasspathTransformer.lambda$transformAll$9(DefaultCachedClasspathTransformer.java:233)
        ... 2 more
Caused by: java.io.IOException: Failed to process the entry 'META-INF/versions/21/org/h2/util/Utils21.class' from '/Users/jeremy/.m2/repository/com/h2database/h2/2.2.224/h2-2.2.224.jar'
        at org.gradle.internal.classpath.InstrumentingClasspathFileTransformer.lambda$visitEntries$1(InstrumentingClasspathFileTransformer.java:161)
        at org.gradle.internal.classpath.ClasspathWalker.visitJarContents(ClasspathWalker.java:91)
        at org.gradle.internal.classpath.ClasspathWalker.visit(ClasspathWalker.java:55)
        at org.gradle.internal.classpath.InstrumentingClasspathFileTransformer.visitEntries(InstrumentingClasspathFileTransformer.java:148)
        at org.gradle.internal.classpath.InstrumentingClasspathFileTransformer.lambda$instrument$0(InstrumentingClasspathFileTransformer.java:139)
        at org.gradle.internal.classpath.ClasspathBuilder.buildJar(ClasspathBuilder.java:68)
        at org.gradle.internal.classpath.ClasspathBuilder.jar(ClasspathBuilder.java:55)
        ... 9 more
Caused by: java.lang.IllegalArgumentException: Unsupported class file major version 65
        at org.objectweb.asm.ClassReader.<init>(ClassReader.java:199)
        at org.objectweb.asm.ClassReader.<init>(ClassReader.java:180)
        at org.objectweb.asm.ClassReader.<init>(ClassReader.java:166)
        at org.gradle.internal.classpath.InstrumentingClasspathFileTransformer.lambda$visitEntries$1(InstrumentingClasspathFileTransformer.java:151)
        ... 15 more


* Get more help at https://help.gradle.org

BUILD FAILED in 1s
```

### Expected Behavior

The execution of the gradle plugin/task should succeed (well, in the case of dependency-check it **might** fail for other reasons). However, a plugin using a multi-release JAR should be able to be executed.

### Context (optional)

This may be related to #24390 and  #24515.

I am the author of the dependency-check plugin which is downloaded from central over 1 million times per month. We upgraded h2 so our user base would no longer ask about CVE-2022-45868.

### Steps to Reproduce

provided above

### Gradle version

7.6.3

### Build scan URL (optional)

_No response_

### Your Environment (optional)

_No response_

This issue is absolutely wrecking us. Some context about our situation:

* We're Palantir, a major customer of Gradle and Gradle Enterprise.
* We're in the middle of an enormous Gradle 7 -> Gradle 8 migration for our internal repos
  * We have over 4000 repos using Gradle 7, and it's taking some time, even with lots of automation, to convert them all
* We also sell a product (Foundry) where Gradle is used by end users. Nearly all these repos are on Gradle 7 and are hard to upgrade.
* We have the US Government as our customer which brings a lot of compliance requirements, as well as other goverments with hard compliance requirements. For the software we ship (including Gradle plugins used by our end users), we _must_ upgrade to a newer version of a library in a very short time frame if it ends up having a CVE.
  * I appreciate this is an "us" problem, but this is unfortunately our reality.

---

[Jackson 2.16.1's `jackson-core` library](https://repo1.maven.org/maven2/com/fasterxml/jackson/core/jackson-core/2.16.1/) is a multi-release jar containing a Java 21 (class file major version 65) class:

```
META-INF/versions/21/com/fasterxml/jackson/core/io/doubleparser/FastDoubleSwar.class
```

When this library is used in a plugin classpath on Gradle 7.6.3, the `asm` library that Gradle has bundled is unable to read this classfile and errors out, failing the build:

<details>
<summary>Stacktrace</summary>

```
Caused by: java.io.IOException: Failed to process the entry 'META-INF/versions/21/com/fasterxml/jackson/core/io/doubleparser/FastDoubleSwar.class' from '/Users/callumr/.gradle/caches/modules-2/files-2.1/com.fasterxml.jackson.core/jackson-core/2.16.1/9456bb3cdd0f79f91a5f730a1b1bb041a380c91f/jackson-core-2.16.1.jar'
	at org.gradle.internal.classpath.InstrumentingClasspathFileTransformer.lambda$visitEntries$1(InstrumentingClasspathFileTransformer.java:161)
	at org.gradle.internal.classpath.ClasspathWalker.visitJarContents(ClasspathWalker.java:91)
	... snip
Caused by: java.lang.IllegalArgumentException: Unsupported class file major version 65
	at org.objectweb.asm.ClassReader.<init>(ClassReader.java:199)
	at org.objectweb.asm.ClassReader.<init>(ClassReader.java:180)
	at org.objectweb.asm.ClassReader.<init>(ClassReader.java:166)
	at org.gradle.internal.classpath.InstrumentingClasspathFileTransformer.lambda$visitEntries$1(InstrumentingClasspathFileTransformer.java:151)
	... 15 more
```

</details>

---

We're in a risky situation: we may need to upgrade to Jackson 2.16.1 at any moment in order to satisfy compliance requirements around security vulnerabilities, but cannot upgrade it due to most repos still being on Gradle 7. We are trying to move to Gradle 8 but it is taking a very long time.

Any sort of backport that helps this situation would be greatly appreciated:
* Bumping `asm` would get us out of the immediate danger and give us more time to complete the Gradle 8 migration.
* Newer Gradle seems to have code around this to fix it, which would be even better:
  * [ClasspathElementTransformForAgent](https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/ClasspathElementTransformForAgent.java#L53-L65)
  * [MultiReleaseClasspathElementTransformForLegacy](https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/MultiReleaseClasspathElementTransformForLegacy.java#L29-L34)

---

<details>
<summary>For completeness, here are all things I have tried to to fix this problem on my end:</summary>

* Write a settings plugin that downgrades Jackson using `resolutionStrategy.eachDependency` on the buildscript's `classpath` configuration.
  * This plugin then needs to be applied to every Gradle build
  * It doesn't work in testkit tests, as they build the plugin classpath out of `file(...)` calls rather than actual dependencies.
  * This doesn't work for some of our uses end-user use cases
* Write a settings plugin that applies an artifact transform to each project's `ScriptHandler` to strip out the multi-release parts of the jars
  * Unfortunately, artifact transforms do not seem to work in `ScriptHandler`s.
  * I think this is because `ScriptHandler`s are pared down minimal versions of `Project`s and don't support artifact transforms (they don't even have any registered `artifactTypes`).
* Attempt to include a higher `asm` in a settings plugin relocated to `org.gradle.internal.impldep.org.objectweb.asm`, hoping that it will trick Gradle into using that instead
  * This plain does not work.

</details>

I think the only reasonable way to resolve this is a backport. Please can you forward this request onto the relevant team for us?

--- a/subprojects/base-services/src/main/java/org/gradle/internal/classanalysis/AsmConstants.java
+++ b/subprojects/base-services/src/main/java/org/gradle/internal/classanalysis/AsmConstants.java
@@ -20,4 +20,13 @@ import org.objectweb.asm.Opcodes;
 
 public class AsmConstants {
     public static final int ASM_LEVEL = Opcodes.ASM9;
+
+    /**
+     * The latest version of Java for which ASM understands the bytecodes.
+     * <p>
+     * Updated for ASM 9.4.
+     *
+     * @see <a href="https://asm.ow2.io/versions.html">ASM release notes</a>
+     */
+    public static final int MAX_SUPPORTED_JAVA_VERSION = 20;
 }

--- a/subprojects/core/src/integTest/groovy/org/gradle/internal/classpath/BuildScriptClasspathIntegrationSpec.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/internal/classpath/BuildScriptClasspathIntegrationSpec.groovy
@@ -16,13 +16,16 @@
 
 package org.gradle.internal.classpath
 
+import org.gradle.api.JavaVersion
 import org.gradle.cache.internal.LeastRecentlyUsedCacheCleanup
 import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.AvailableJavaHomes
 import org.gradle.integtests.fixtures.cache.FileAccessTimeJournalFixture
 import org.gradle.integtests.fixtures.executer.ArtifactBuilder
 import org.gradle.test.fixtures.file.TestFile
 import org.gradle.test.fixtures.server.http.HttpServer
 import org.gradle.test.fixtures.server.http.MavenHttpRepository
+import org.gradle.util.Requires
 import org.junit.Rule
 import spock.lang.Issue
 import spock.lang.Unroll
@@ -310,6 +313,83 @@ class BuildScriptClasspathIntegrationSpec extends AbstractIntegrationSpec implem
 
         then:
         noExceptionThrown()
+    }
+
+    @Requires(adhoc = {
+        AvailableJavaHomes.getJdk(JavaVersion.toVersion(8)) && AvailableJavaHomes.getJdk(JavaVersion.toVersion(11))
+    })
+    def "proper version is selected for multi-release jar"() {
+        given:
+        createDir("mrjar") {
+            file("build.gradle") << """
+                plugins {
+                    id("java")
+                    id 'me.champeau.mrjar' version "0.1.1"
+                }
+
+                multiRelease {
+                    targetVersions 8, 11
+                }
+            """
+            file("src/main/java/org/gradle/test/mrjar/Foo.java") << """
+                package org.gradle.test.mrjar;
+
+                public class Foo {
+                    public static String getBar() {
+                        return "DEFAULT";
+                    }
+                }
+            """
+            file("src/java11/java/org/gradle/test/mrjar/Foo.java") << """
+                package org.gradle.test.mrjar;
+
+                public class Foo {
+                    public static String getBar() {
+                        return "11";
+                    }
+                }
+            """
+        }
+
+        buildScript("""
+            buildscript {
+                dependencies {
+                    classpath "org.gradle.test:mrjar:1.+"
+                }
+            }
+
+            import org.gradle.test.mrjar.Foo
+
+            tasks.register("printFoo") {
+                doLast {
+                    println("JAR = \${Foo.bar}")
+                }
+            }
+        """)
+        settingsFile("""
+            includeBuild("mrjar") {
+                dependencySubstitution {
+                    substitute module('org.gradle.test:mrjar') using project(':')
+                }
+            }
+        """)
+
+        def java8Home = AvailableJavaHomes.getJdk8().javaHome
+        def java11Home = AvailableJavaHomes.getJdk11().javaHome
+
+        when:
+        executer.withJavaHome(java8Home).withArguments("-Porg.gradle.java.installations.paths=$java8Home,$java11Home")
+        succeeds("printFoo")
+
+        then:
+        outputContains("JAR = DEFAULT")
+
+        when:
+        executer.withJavaHome(java11Home).withArguments("-Porg.gradle.java.installations.paths=$java8Home,$java11Home")
+        succeeds("printFoo")
+
+        then:
+        outputContains("JAR = 11")
     }
 
     void notInJarCache(String filename) {

--- a/subprojects/core/src/main/java/org/gradle/internal/classpath/InstrumentingClasspathFileTransformer.java
+++ b/subprojects/core/src/main/java/org/gradle/internal/classpath/InstrumentingClasspathFileTransformer.java
@@ -24,6 +24,7 @@ import org.gradle.api.internal.file.archive.impl.FileZipInput;
 import org.gradle.cache.FileLock;
 import org.gradle.cache.FileLockManager;
 import org.gradle.internal.Pair;
+import org.gradle.internal.classanalysis.AsmConstants;
 import org.gradle.internal.file.FileException;
 import org.gradle.internal.file.FileType;
 import org.gradle.internal.hash.HashCode;
@@ -31,6 +32,7 @@ import org.gradle.internal.hash.Hasher;
 import org.gradle.internal.hash.Hashing;
 import org.gradle.internal.snapshot.FileSystemLocationSnapshot;
 import org.gradle.util.internal.GFileUtils;
+import org.gradle.util.internal.JarUtil;
 import org.objectweb.asm.ClassReader;
 import org.objectweb.asm.ClassVisitor;
 import org.objectweb.asm.ClassWriter;
@@ -39,6 +41,7 @@ import org.slf4j.LoggerFactory;
 
 import java.io.File;
 import java.io.IOException;
+import java.util.OptionalInt;
 
 import static java.lang.String.format;
 import static org.gradle.cache.internal.filelock.LockOptionsBuilder.mode;
@@ -52,6 +55,18 @@ class InstrumentingClasspathFileTransformer implements ClasspathFileTransformer 
     private final ClasspathBuilder classpathBuilder;
     private final CachedClasspathTransformer.Transform transform;
     private final HashCode configHash;
+
+    /**
+     * A "pending" transformation of the original file/directory.
+     */
+    private interface Transformation {
+        /**
+         * Transform the file/directory into destination. The destination should be a JAR file name.
+         *
+         * @param destination the destination file
+         */
+        void transform(File destination);
+    }
 
     public InstrumentingClasspathFileTransformer(
         FileLockManager fileLockManager,
@@ -69,6 +84,7 @@ class InstrumentingClasspathFileTransformer implements ClasspathFileTransformer 
     private HashCode configHashFor(CachedClasspathTransformer.Transform transform) {
         Hasher hasher = Hashing.defaultFunction().newHasher();
         hasher.putInt(CACHE_FORMAT);
+        hasher.putInt(AsmConstants.MAX_SUPPORTED_JAVA_VERSION);
         transform.applyConfigurationTo(hasher);
         return hasher.hash();
     }
@@ -125,60 +141,180 @@ class InstrumentingClasspathFileTransformer implements ClasspathFileTransformer 
     }
 
     private void transform(File source, File dest) {
-        if (isSignedJar(source)) {
+        createTransformer(source).transform(dest);
+    }
+
+    private Transformation createTransformer(File source) {
+        Boolean isMultiReleaseJar = null;
+
+        if (source.isFile()) {
+            // Walk a file to figure out if it is signed and if it is a multi-release JAR.
+            try (ZipInput entries = FileZipInput.create(source)) {
+                for (ZipEntry entry : entries) {
+                    String entryName = entry.getName();
+                    if (isJarSignatureFile(entryName)) {
+                        // TODO(mlopatkin) Manifest of the signed JAR contains signature information and must be the first entry in the JAR.
+                        //  Looking into the manifest here should be more effective.
+                        // This policy doesn't transform signed JARs so no further checks are necessary.
+                        return new SkipTransformation(source);
+                    }
+                    if (isMultiReleaseJar == null && JarUtil.isManifestName(entryName)) {
+                        isMultiReleaseJar = JarUtil.isMultiReleaseJarManifest(JarUtil.readManifest(entry.getContent()));
+                    }
+                }
+            } catch (FileException e) {
+                // Ignore malformed archive, let the transformation handle it.
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }
+        return new InstrumentingTransformation(source, isMultiReleaseJar != null && isMultiReleaseJar);
+    }
+
+    private boolean isJarSignatureFile(String entryName) {
+        return entryName.startsWith("META-INF/") && entryName.endsWith(".SF");
+    }
+
+    /**
+     * A no-op transformation that copies the original file verbatim. Can be used if the original cannot be instrumented under policy.
+     */
+    private static class SkipTransformation implements Transformation {
+        private final File source;
+
+        public SkipTransformation(File source) {
+            this.source = source;
+        }
+
+        @Override
+        public void transform(File destination) {
             LOGGER.debug("Signed archive '{}'. Skipping instrumentation.", source.getName());
-            GFileUtils.copyFile(source, dest);
-        } else {
-            instrument(source, dest);
+            GFileUtils.copyFile(source, destination);
         }
     }
 
-    private void instrument(File source, File dest) {
-        classpathBuilder.jar(dest, builder -> {
-            try {
-                visitEntries(source, builder);
-            } catch (FileException e) {
-                // Badly formed archive, so discard the contents and produce an empty JAR
-                LOGGER.debug("Malformed archive '{}'. Discarding contents.", source.getName(), e);
-            }
-        });
-    }
+    /**
+     * Base class for the transformations. Note that the order in which entries are visited is not defined.
+     */
+    private class InstrumentingTransformation implements Transformation {
+        protected final File source;
+        private final boolean isMultiReleaseJar;
 
-    private void visitEntries(File source, ClasspathBuilder.EntryBuilder builder) throws IOException, FileException {
-        classpathWalker.visit(source, entry -> {
+        public InstrumentingTransformation(File source, boolean isMultiReleaseJar) {
+            this.source = source;
+            this.isMultiReleaseJar = isMultiReleaseJar;
+        }
+
+        @Override
+        public final void transform(File destination) {
+            classpathBuilder.jar(destination, builder -> {
+                try {
+                    visitEntries(builder);
+                } catch (FileException e) {
+                    // Badly formed archive, so discard the contents and produce an empty JAR
+                    LOGGER.debug("Malformed archive '{}'. Discarding contents.", source.getName(), e);
+                }
+            });
+        }
+
+        private void visitEntries(ClasspathBuilder.EntryBuilder builder) throws IOException, FileException {
+            classpathWalker.visit(source, entry -> {
+                visitEntry(builder, entry);
+            });
+        }
+
+        private void visitEntry(ClasspathBuilder.EntryBuilder builder, ClasspathEntryVisitor.Entry entry) throws IOException {
             try {
-                if (entry.getName().endsWith(".class")) {
-                    ClassReader reader = new ClassReader(entry.getContent());
-                    ClassWriter classWriter = new ClassWriter(ClassWriter.COMPUTE_MAXS);
-                    Pair<RelativePath, ClassVisitor> chain = transform.apply(entry, classWriter);
-                    reader.accept(chain.right, 0);
-                    byte[] bytes = classWriter.toByteArray();
-                    builder.put(chain.left.getPathString(), bytes, entry.getCompressionMethod());
+                if (isClassFile(entry)) {
+                    processClassFile(builder, entry);
+                } else if (isManifest(entry)) {
+                    processManifest(builder, entry);
                 } else {
-                    builder.put(entry.getName(), entry.getContent(), entry.getCompressionMethod());
+                    processResource(builder, entry);
                 }
             } catch (Throwable e) {
                 throw new IOException("Failed to process the entry '" + entry.getName() + "' from '" + source + "'", e);
             }
-        });
-    }
+        }
 
-    private boolean isSignedJar(File source) {
-        if (!source.isFile()) {
+        /**
+         * Processes a class file. The type of file is determined solely by name, so it may not be a well-formed class file.
+         * Base class implementation applies the {@link InstrumentingClasspathFileTransformer#transform} to the code.
+         *
+         * @param builder the builder for the transformed output
+         * @param classEntry the entry to process
+         * @throws IOException if reading or writing entry fails
+         */
+        protected void processClassFile(ClasspathBuilder.EntryBuilder builder, ClasspathEntryVisitor.Entry classEntry) throws IOException {
+            if (isMultiReleaseJar && isInUnsupportedMrJarVersionedDirectory(classEntry)) {
+                // The entries should only be filtered out if we're transforming the proper multi-release JAR.
+                // Otherwise, even if the entry path looks like it is inside the versioned directory, it may still be accessed as a
+                // resource.
+                // Of course, user code can try to access resources inside versioned directories with full paths anyway, but that's
+                // a tradeoff we're making.
+                return;
+            }
+
+            ClassReader reader = new ClassReader(classEntry.getContent());
+            ClassWriter classWriter = new ClassWriter(ClassWriter.COMPUTE_MAXS);
+            Pair<RelativePath, ClassVisitor> chain = transform.apply(classEntry, classWriter);
+            reader.accept(chain.right, 0);
+            byte[] bytes = classWriter.toByteArray();
+            builder.put(chain.left.getPathString(), bytes, classEntry.getCompressionMethod());
+        }
+
+        /**
+         * Processes a JAR Manifest. Base class implementation copies the manifest unchanged.
+         *
+         * @param builder the builder for the transformed output
+         * @param manifestEntry the entry to process
+         * @throws IOException if reading or writing entry fails
+         */
+        protected void processManifest(ClasspathBuilder.EntryBuilder builder, ClasspathEntryVisitor.Entry manifestEntry) throws IOException {
+            copyResource(builder, manifestEntry);
+        }
+
+        /**
+         * Processes a resource entry. Base class implementation copies the resource unchanged.
+         *
+         * @param builder the builder for the transformed output
+         * @param resourceEntry the entry to process
+         * @throws IOException if reading or writing entry fails
+         */
+        protected void processResource(ClasspathBuilder.EntryBuilder builder, ClasspathEntryVisitor.Entry resourceEntry) throws IOException {
+            if (isMultiReleaseJar && isInUnsupportedMrJarVersionedDirectory(resourceEntry)) {
+                // The entries should only be filtered out if we're transforming the proper multi-release JAR.
+                // Otherwise, even if the entry path looks like it is inside the versioned directory, it may still be accessed as a
+                // resource.
+                // Of course, user code can try to access resources inside versioned directories with full paths anyway, but that's
+                // a tradeoff we're making.
+                return;
+            }
+            copyResource(builder, resourceEntry);
+        }
+
+        private void copyResource(ClasspathBuilder.EntryBuilder builder, ClasspathEntryVisitor.Entry resourceEntry) throws IOException {
+            builder.put(resourceEntry.getName(), resourceEntry.getContent(), resourceEntry.getCompressionMethod());
+        }
+
+        private boolean isClassFile(ClasspathEntryVisitor.Entry entry) {
+            return entry.getName().endsWith(".class");
+        }
+
+        private boolean isManifest(ClasspathEntryVisitor.Entry entry) {
+            return JarUtil.isManifestName(entry.getName());
+        }
+
+        private boolean isInUnsupportedMrJarVersionedDirectory(ClasspathEntryVisitor.Entry entry) {
+            OptionalInt version = JarUtil.getVersionedDirectoryMajorVersion(entry.getName());
+            if (version.isPresent()) {
+                return !isSupportedVersion(version.getAsInt());
+            }
+            // The entry is not in the versioned directory at all.
             return false;
         }
-        try (ZipInput entries = FileZipInput.create(source)) {
-            for (ZipEntry entry : entries) {
-                String entryName = entry.getName();
-                if (entryName.startsWith("META-INF/") && entryName.endsWith(".SF")) {
-                    return true;
-                }
-            }
-        } catch (FileException e) {
-            // Ignore malformed archive
-        } catch (IOException e) {
-            throw new UncheckedIOException(e);
+
+        private boolean isSupportedVersion(int javaMajorVersion) {
+            return javaMajorVersion <= AsmConstants.MAX_SUPPORTED_JAVA_VERSION;
         }
-        return false;
     }
 }

--- a/subprojects/core/src/main/java/org/gradle/util/internal/JarUtil.java
+++ b/subprojects/core/src/main/java/org/gradle/util/internal/JarUtil.java
@@ -19,11 +19,23 @@ package org.gradle.util.internal;
 import org.apache.commons.io.IOUtils;
 import org.gradle.internal.IoActions;
 
+import javax.annotation.Nullable;
 import java.io.*;
+import java.util.OptionalInt;
+import java.util.jar.JarFile;
+import java.util.jar.Manifest;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 import java.util.zip.ZipEntry;
 import java.util.zip.ZipInputStream;
 
 public class JarUtil {
+    // We cannot use Attributes.Name.MULTI_RELEASE as it is only available since Java 9;
+    public static final String MULTI_RELEASE_ATTRIBUTE = "Multi-Release";
+    // Pattern for JAR entries in the versioned directories.
+    // See https://docs.oracle.com/en/java/javase/20/docs/specs/jar/jar.html#multi-release-jar-files
+    private static final Pattern VERSIONED_JAR_ENTRY_PATH = Pattern.compile("^META-INF/versions/(\\d+)/.+$");
+
     public static boolean extractZipEntry(File jarFile, String entryName, File extractToFile) throws IOException {
         boolean entryExtracted = false;
 
@@ -54,5 +66,65 @@ public class JarUtil {
         }
 
         return entryExtracted;
+    }
+
+    /**
+     * Checks if the {@code jarEntryName} is the name of the JAR manifest entry.
+     *
+     * @param jarEntryName the name of the entry
+     * @return {@code true} if the entry is the JAR manifest
+     */
+    public static boolean isManifestName(String jarEntryName) {
+        return jarEntryName.equals(JarFile.MANIFEST_NAME);
+    }
+
+    /**
+     * Parses Manifest from a byte array.
+     * @param content the bytes of the manifest
+     * @return the Manifest
+     * @throws IOException if the manifest cannot be parsed
+     */
+    public static Manifest readManifest(byte[] content) throws IOException {
+        return new Manifest(new ByteArrayInputStream(content));
+    }
+
+    /**
+     * Checks if the manifest declares JAR to be multi-release.
+     *
+     * @param manifest the manifest
+     * @return {@code true} if the manifest declares the multi-release JAR
+     */
+    public static boolean isMultiReleaseJarManifest(Manifest manifest) {
+        return Boolean.parseBoolean(getManifestMainAttribute(manifest, MULTI_RELEASE_ATTRIBUTE));
+    }
+
+    @Nullable
+    private static String getManifestMainAttribute(Manifest manifest, String name) {
+        return manifest.getMainAttributes().getValue(name);
+    }
+
+    /**
+     * Checks if the entry path is inside a versioned directory and returns the corresponding major version of Java platform.
+     * Returns empty Optional if this entry isn't inside a versioned directory.
+     *
+     * @param entryPath the full path to the entry inside the JAR
+     * @return major version of Java platform for which this entry is intended if it is in the versioned directory or empty Optional
+     *
+     * @see <a href="https://docs.oracle.com/en/java/javase/20/docs/specs/jar/jar.html#multi-release-jar-files">MR JAR specification</a>
+     */
+    public static OptionalInt getVersionedDirectoryMajorVersion(String entryPath) {
+        Matcher match = VERSIONED_JAR_ENTRY_PATH.matcher(entryPath);
+        if (match.matches()) {
+            try {
+                int version = Integer.parseInt(match.group(1));
+                return OptionalInt.of(version);
+            } catch (NumberFormatException ignored) {
+                // Even though the pattern ensures that the version name is all digits, it fails to parse, probably because it is too big.
+                // Technically it may be a valid MR JAR for Java >Integer.MAX_VALUE, but we are too far away from this.
+                // We assume that JAR author didn't intend it to be a versioned directory.
+            }
+        }
+        // The entry is not in the versioned directory.
+        return OptionalInt.empty();
     }
 }

--- a/subprojects/core/src/test/groovy/org/gradle/internal/classpath/DefaultCachedClasspathTransformerTest.groovy
+++ b/subprojects/core/src/test/groovy/org/gradle/internal/classpath/DefaultCachedClasspathTransformerTest.groovy
@@ -238,7 +238,7 @@ class DefaultCachedClasspathTransformerTest extends ConcurrentSpec {
         def file = testDir.file("thing.jar")
         jar(file)
         def classpath = DefaultClassPath.of(file)
-        def cachedFile = testDir.file("cached/16fa923fc1ec782f28887195acdc5401/thing.jar")
+        def cachedFile = testDir.file("cached/7b0ef65a9775af8184c3570e1953ea12/thing.jar")
 
         when:
         def cachedClasspath = transformer.transform(classpath, BuildLogic)
@@ -266,7 +266,7 @@ class DefaultCachedClasspathTransformerTest extends ConcurrentSpec {
         def dir = testDir.file("thing.dir")
         classesDir(dir)
         def classpath = DefaultClassPath.of(dir)
-        def cachedFile = testDir.file("cached/d4d760260e067b91216f90d0d5a7ac5c/thing.dir.jar")
+        def cachedFile = testDir.file("cached/0a622a9ece61fd0075c0585af445556d/thing.dir.jar")
 
         when:
         def cachedClasspath = transformer.transform(classpath, BuildLogic)
@@ -296,8 +296,8 @@ class DefaultCachedClasspathTransformerTest extends ConcurrentSpec {
         def file = testDir.file("thing.jar")
         jar(file)
         def classpath = DefaultClassPath.of(dir, file)
-        def cachedDir = testDir.file("cached/d4d760260e067b91216f90d0d5a7ac5c/thing.dir.jar")
-        def cachedFile = testDir.file("cached/16fa923fc1ec782f28887195acdc5401/thing.jar")
+        def cachedDir = testDir.file("cached/0a622a9ece61fd0075c0585af445556d/thing.dir.jar")
+        def cachedFile = testDir.file("cached/7b0ef65a9775af8184c3570e1953ea12/thing.jar")
 
         when:
         def cachedClasspath = transformer.transform(classpath, BuildLogic)
@@ -354,8 +354,8 @@ class DefaultCachedClasspathTransformerTest extends ConcurrentSpec {
         def file3 = testDir.file("thing3.jar")
         jar(file3)
         def classpath = DefaultClassPath.of(dir, file, dir2, file2, dir3, file3)
-        def cachedDir = testDir.file("cached/d4d760260e067b91216f90d0d5a7ac5c/thing.dir.jar")
-        def cachedFile = testDir.file("cached/16fa923fc1ec782f28887195acdc5401/thing.jar")
+        def cachedDir = testDir.file("cached/0a622a9ece61fd0075c0585af445556d/thing.dir.jar")
+        def cachedFile = testDir.file("cached/7b0ef65a9775af8184c3570e1953ea12/thing.jar")
 
         when:
         def cachedClasspath = transformer.transform(classpath, BuildLogic)
@@ -375,7 +375,7 @@ class DefaultCachedClasspathTransformerTest extends ConcurrentSpec {
         def file = testDir.file("thing.jar")
         jar(file)
         def classpath = DefaultClassPath.of(file)
-        def cachedFile = testDir.file("cached/90b54e71cfa1cf2192b11e3e12e93caa/thing.jar")
+        def cachedFile = testDir.file("cached/b54a095af3ca71c47824868b12d5cdfb/thing.jar")
 
         when:
         def cachedClasspath = transformer.transform(classpath, BuildLogic, transform)
@@ -409,7 +409,7 @@ class DefaultCachedClasspathTransformerTest extends ConcurrentSpec {
         def file = testDir.file("thing.jar")
         jarWithStoredResource(file)
         def classpath = DefaultClassPath.of(file)
-        def cachedFile = testDir.file("cached/40ba77bb7fbecc72eec8f0e8257e55d5/thing.jar")
+        def cachedFile = testDir.file("cached/ced05151ad737d8676bc78ca5914f46a/thing.jar")
 
         when:
         def cachedClasspath = transformer.transform(classpath, BuildLogic)

--- a/subprojects/core/src/test/groovy/org/gradle/internal/classpath/InstrumentingClasspathFileTransformerTest.groovy
+++ b/subprojects/core/src/test/groovy/org/gradle/internal/classpath/InstrumentingClasspathFileTransformerTest.groovy
@@ -1,0 +1,204 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.classpath
+
+import org.gradle.api.file.RelativePath
+import org.gradle.api.internal.file.TestFiles
+import org.gradle.cache.FileLockManager
+import org.gradle.internal.MutableReference
+import org.gradle.internal.Pair
+import org.gradle.internal.classanalysis.AsmConstants
+import org.gradle.internal.hash.Hasher
+import org.gradle.internal.snapshot.FileSystemLocationSnapshot
+import org.gradle.test.fixtures.archive.JarTestFixture
+import org.gradle.test.fixtures.file.TestNameTestDirectoryProvider
+import org.junit.Rule
+import org.objectweb.asm.ClassVisitor
+import spock.lang.Specification
+
+import java.util.jar.JarFile
+
+import static org.gradle.util.JarUtils.jar
+
+class InstrumentingClasspathFileTransformerTest extends Specification {
+    @Rule
+    TestNameTestDirectoryProvider testDirectoryProvider = new TestNameTestDirectoryProvider(getClass())
+    def testDir = testDirectoryProvider.testDirectory
+
+    def cacheDir = testDir.createDir("cached")
+    def classpathWalker = new ClasspathWalker(TestFiles.fileSystem())
+    def classpathBuilder = new ClasspathBuilder(TestFiles.tmpDirTemporaryFileProvider(testDirectoryProvider.createDir("tmp")))
+    def fileLockManager = Stub(FileLockManager)
+    def fileSystemAccess = TestFiles.fileSystemAccess()
+
+    def "instrumentation preserves classes"() {
+        given:
+        def testFile = jar(testDir.file("thing.jar")) {
+            manifest {}
+            entry("Foo.class", classOne())
+        }
+
+        expect:
+        with(jarFixture(transform(testFile))) {
+            assertContainsFile("Foo.class")
+        }
+    }
+
+    def "instrumentation copies manifest into transformed jar"() {
+        given:
+        def testFile = jar(testDir.file("thing.jar")) {
+            manifest {
+                mainAttributes.putValue("Some-Value", "theValue")
+            }
+            entry("Foo.class", classOne())
+        }
+
+        expect:
+        with(jarFixture(transform(testFile))) {
+            assertContainsFile(JarFile.MANIFEST_NAME)
+
+            assert manifest.mainAttributes.getValue("Some-Value") == "theValue"
+        }
+    }
+
+    def "instrumentation produces multi-release jar"() {
+        given:
+        def testFile = jar(testDir.file("thing.jar")) {
+            manifest {
+                mainAttributes.putValue("Multi-Release", "true")
+            }
+
+            entry("Foo.class", classOne())
+            versionedEntry(AsmConstants.MAX_SUPPORTED_JAVA_VERSION, "Foo.class", classOne())
+        }
+
+        expect:
+        with(jarFixture(transform(testFile))) {
+            assertIsMultiRelease()
+            assertContainsFile("Foo.class")
+            assertContainsVersioned(AsmConstants.MAX_SUPPORTED_JAVA_VERSION, "Foo.class")
+        }
+    }
+
+    def "instrumentation removes unsupported versioned directories from multi-release jar"() {
+        given:
+        def testFile = jar(testDir.file("thing.jar")) {
+            manifest {
+                mainAttributes.putValue("Multi-Release", "true")
+            }
+
+            entry("Foo.class", classOne())
+            versionedEntry(AsmConstants.MAX_SUPPORTED_JAVA_VERSION + 1, "Foo.class", classOne())
+        }
+
+        expect:
+        with(jarFixture(transform(testFile))) {
+            assertContainsFile("Foo.class")
+            assertNotContainsVersioned(AsmConstants.MAX_SUPPORTED_JAVA_VERSION, "Foo.class")
+        }
+    }
+
+    def "instrumentation copies resources into transformed jar"() {
+        given:
+        def testFile = jar(testDir.file("thing.jar")) {
+            manifest {}
+
+            entry("Foo.class", classOne())
+            entry("resource.txt", "resource body")
+        }
+
+        expect:
+        with(jarFixture(transform(testFile))) {
+            assertFileContent("resource.txt", "resource body")
+        }
+    }
+
+    def "instrumentation removes unsupported versioned resources from transformed jar"() {
+        given:
+        def testFile = jar(testDir.file("thing.jar")) {
+            manifest {
+                mainAttributes.putValue("Multi-Release", "true")
+            }
+
+            entry("Foo.class", classOne())
+            entry("resource.txt", "resource body")
+            versionedEntry(AsmConstants.MAX_SUPPORTED_JAVA_VERSION, "resource.txt", "resource MAX body")
+            versionedEntry(AsmConstants.MAX_SUPPORTED_JAVA_VERSION + 1, "resource.txt", "resource unsupported body")
+        }
+
+        expect:
+        with(jarFixture(transform(testFile))) {
+            assertContainsVersioned(AsmConstants.MAX_SUPPORTED_JAVA_VERSION, "resource.txt")
+            assertNotContainsVersioned(AsmConstants.MAX_SUPPORTED_JAVA_VERSION + 1, "resource.txt")
+        }
+    }
+
+    def "multi-release jar must have manifest attribute to be processed by instrumentation"() {
+        given:
+        def testFile = jar(testDir.file("thing.jar")) {
+            manifest {
+                // No Multi-Release attribute in the manifest.
+            }
+
+            entry("Foo.class", classOne())
+            versionedEntry(AsmConstants.MAX_SUPPORTED_JAVA_VERSION + 1, "Foo.class", classOne())
+        }
+
+        expect:
+        with(jarFixture(transform(testFile))) {
+            assertContainsVersioned(AsmConstants.MAX_SUPPORTED_JAVA_VERSION + 1, "Foo.class")
+        }
+    }
+
+
+    private File transform(File file) {
+        def transformer = new InstrumentingClasspathFileTransformer(
+            fileLockManager,
+            classpathWalker,
+            classpathBuilder,
+            new NoOpTransformer()
+        )
+        return transformer.transform(file, snapshot(file), cacheDir)
+    }
+
+
+    private static class NoOpTransformer implements CachedClasspathTransformer.Transform {
+        @Override
+        void applyConfigurationTo(Hasher hasher) {
+        }
+
+        @Override
+        Pair<RelativePath, ClassVisitor> apply(ClasspathEntryVisitor.Entry entry, ClassVisitor visitor) {
+            return Pair.of(entry.path, visitor)
+        }
+    }
+
+    byte[] classOne() {
+        return getClass().classLoader.getResource(SystemPropertyAccessingThing.name.replace('.', '/') + ".class").bytes
+    }
+
+    private static JarTestFixture jarFixture(File transformedJar) {
+        def expectManifest = true
+        return new JarTestFixture(transformedJar, 'UTF-8', null, expectManifest)
+    }
+
+    private FileSystemLocationSnapshot snapshot(File file) {
+        MutableReference<FileSystemLocationSnapshot> result = MutableReference.empty()
+        fileSystemAccess.read(file.getAbsolutePath(), result.&set)
+        return result.get()
+    }
+}

--- a/subprojects/core/src/testFixtures/groovy/org/gradle/util/JarUtils.groovy
+++ b/subprojects/core/src/testFixtures/groovy/org/gradle/util/JarUtils.groovy
@@ -16,7 +16,17 @@
 
 package org.gradle.util
 
+import org.gradle.test.fixtures.archive.JarTestFixture
+import org.gradle.util.internal.JarUtil
+
+import java.nio.charset.StandardCharsets
+import java.time.LocalDateTime
+import java.time.ZoneId
+import java.util.jar.Attributes
+import java.util.jar.JarEntry
+import java.util.jar.JarFile
 import java.util.jar.JarOutputStream
+import java.util.jar.Manifest
 import java.util.zip.ZipEntry
 
 class JarUtils {
@@ -34,5 +44,140 @@ class JarUtils {
             jarOut.close()
         }
         return out.toByteArray()
+    }
+
+    /**
+     * Builds a JAR file with a DSL. The JAR file uses deterministic timestamps. The Jar manifest must be configured explicitly, as a first step.
+     * <p>
+     * <pre>
+     *     jar("path/to/file.jar") {
+     *         manifest {  // can be withoutManifest()
+     *             mainAttributes.putValue("Multi-Release", "true")
+     *         }
+     *
+     *         entry("Foo.class", fooBytes)
+     *         versionedEntry(11, "Foo.class", fooBytesForJava11)
+     *     }
+     *
+     * </pre>
+     * @param jarFile the jar file to create
+     * @param closure the configuration of the JAR file
+     * @return the path to the created JAR file.
+     */
+    static File jar(File jarFile, @DelegatesTo(value = JarBuilder, strategy = Closure.DELEGATE_FIRST) Closure<?> closure) {
+        try (def out = jarFile.newOutputStream(); def jarOut = new JarOutputStream(out)) {
+            closure.setDelegate(new JarBuilder(jarOut))
+            closure()
+        }
+
+        jarFile
+    }
+
+    /**
+     * DSL helper to build the JAR file. Start building with configuring the manifest ({@link JarBuilder#manifest(groovy.lang.Closure)}
+     * or specifying the lack thereof ({@link JarBuilder#withoutManifest()}.
+     */
+    static class JarBuilder {
+        /**
+         * @see org.gradle.api.internal.file.archive.ZipCopyAction#CONSTANT_TIME_FOR_ZIP_ENTRIES
+         */
+        private static final long CONSTANT_TIME_FOR_ZIP_ENTRIES = LocalDateTime.parse("1980-02-01T00:00:00.00").atZone(ZoneId.systemDefault()).toInstant().toEpochMilli()
+
+        private final JarOutputStream jarOut
+        private boolean hasConfiguredManifest
+
+        JarBuilder(JarOutputStream jarOut) {
+            this.jarOut = jarOut
+        }
+
+        /**
+         * Specifies that this JAR has no manifest.
+         */
+        void withoutManifest() {
+            assert !hasConfiguredManifest: "Already written the manifest"
+            hasConfiguredManifest = true
+        }
+
+        /**
+         * Configures the JAR manifest.
+         */
+        void manifest(@DelegatesTo(value = ManifestWithDsl, strategy = Closure.DELEGATE_FIRST) Closure<?> closure) {
+            assert !hasConfiguredManifest: "Already written the manifest"
+
+            def man = new ManifestWithDsl()
+            man.mainAttributes.put(Attributes.Name.MANIFEST_VERSION, "1.0")
+
+            closure.setDelegate(man)
+            closure()
+
+            jarOut.putNextEntry(newEntryWithFixedTime(JarFile.MANIFEST_NAME))
+            man.write(jarOut)
+
+            hasConfiguredManifest = true
+            this
+        }
+
+        private static JarEntry newEntryWithFixedTime(String path) {
+            return new JarEntry(path).tap {
+                // setTimeLocal() is better but it is Java 9+.
+                setTime(CONSTANT_TIME_FOR_ZIP_ENTRIES)
+            }
+        }
+
+        /**
+         * Writes a versioned flavor of a JAR entry. The {@code path} should be a non-versioned path, e.g. {@code path/to/Foo.class}.
+         * @param version the java version (9, 10, etc.), must be &gt;8
+         * @param path the path of the entry (non-versioned)
+         * @param bytes the body of the entry
+         */
+        void versionedEntry(int version, String path, byte[] bytes) {
+            entry(JarTestFixture.toVersionedPath(version, path), bytes)
+        }
+
+        /**
+         * Writes a versioned flavor of a JAR entry. The {@code path} should be a non-versioned path, e.g. {@code path/to/Foo.class}.
+         * @param version the java version (9, 10, etc.), must be &gt;8
+         * @param path the path of the entry (non-versioned)
+         * @param body the body of the entry which is encoded with UTF-8
+         */
+        void versionedEntry(int version, String path, String body) {
+            entry(JarTestFixture.toVersionedPath(version, path), body)
+        }
+
+        /**
+         * Writes a JAR entry.
+         * @param path the path of the entry
+         * @param bytes the body of the entry
+         */
+        void entry(String path, byte[] bytes) {
+            checkManifestWritten()
+            jarOut.putNextEntry(newEntryWithFixedTime(path))
+            jarOut.write(bytes)
+        }
+
+        /**
+         * Writes a JAR entry.
+         * @param path the path of the entry
+         * @param body the body of the entry which is encoded with UTF-8
+         */
+        void entry(String path, String body) {
+            entry(path, body.getBytes(StandardCharsets.UTF_8))
+        }
+
+        private void checkManifestWritten() {
+            assert hasConfiguredManifest: "Must have manifest before entries. Use withoutManifest() to skip it explicitly."
+        }
+    }
+
+    /**
+     * A Manifest with additional goodies to make its configuration more pleasant.
+     */
+    static class ManifestWithDsl extends Manifest {
+        /**
+         * Configures this Manifest to be Multi-Release.
+         */
+        void multiRelease() {
+            mainAttributes.putValue(JarUtil.MULTI_RELEASE_ATTRIBUTE, "true")
+        }
     }
 }

--- a/subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/archive/JarTestFixture.groovy
+++ b/subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/archive/JarTestFixture.groovy
@@ -20,21 +20,31 @@ import org.apache.commons.io.IOUtils
 import org.apache.tools.zip.ZipFile
 import org.gradle.test.fixtures.file.ClassFile
 
+import javax.annotation.Nullable
 import java.util.jar.JarEntry
 import java.util.jar.JarFile
 import java.util.jar.Manifest
 
 class JarTestFixture extends ZipTestFixture {
     File file
+    private @Nullable Manifest theManifest
+    /**
+     * Asserts that the Jar file is well-formed.
+     */
+    JarTestFixture(File file, String metadataCharset = 'UTF-8', String contentCharset = null) {
+        this(file, metadataCharset, contentCharset, true)
+    }
 
     /**
-     * Asserts that the Jar file is well-formed
+     * Creates the fixture.
      */
-     JarTestFixture(File file, String metadataCharset = 'UTF-8', String contentCharset = null) {
-         super(file, metadataCharset, contentCharset)
-         this.file = file
-         isManifestPresentAndFirstEntry()
-     }
+    JarTestFixture(File file, String metadataCharset, String contentCharset, boolean checkManifest) {
+        super(file, metadataCharset, contentCharset)
+        this.file = file
+        if (checkManifest) {
+            isManifestPresentAndFirstEntry()
+        }
+    }
 
     /**
      * Asserts that the given service is defined in this jar file.
@@ -83,6 +93,31 @@ class JarTestFixture extends ZipTestFixture {
     }
 
     Manifest getManifest() {
-        new Manifest(IOUtils.toInputStream(content('META-INF/MANIFEST.MF'), contentCharset))
+        if (theManifest == null) {
+            theManifest = readManifest()
+        }
+        return theManifest
+    }
+
+    private Manifest readManifest() {
+        return new Manifest(IOUtils.toInputStream(content('META-INF/MANIFEST.MF'), contentCharset as String))
+    }
+
+    def assertIsMultiRelease() {
+        assert "true" == manifest.mainAttributes.getValue("Multi-Release")
+        this
+    }
+
+    def assertContainsVersioned(int version, String basePath) {
+        assertContainsFile(toVersionedPath(version, basePath))
+    }
+
+    def assertNotContainsVersioned(int version, String basePath) {
+        assertNotContainsFile(toVersionedPath(version, basePath))
+    }
+
+    static String toVersionedPath(int version, String basePath) {
+        assert version > 8: "Java only supports versioned directories for versions >8, got $version"
+        return "META-INF/versions/$version/$basePath"
     }
 }

